google/cloud/bigtable/_mutate_rows.py

```diff
@@ -0,0 +1,220 @@
+from google.cloud.bigtable._helpers import _attempt_timeout_generator
+
+if TYPE_CHECKING:
+    from google.cloud.bigtable_v2.services.bigtable.async_client import (
+        BigtableAsyncClient,
+    )
+    from google.cloud.bigtable.client import Table
+    from google.cloud.bigtable.mutations import RowMutationEntry
+    from google.cloud.bigtable_v2.types.bigtable import MutateRowsResponse
+
+
+class _MutateRowsIncomplete(RuntimeError):
+    """
+    Exception raised when a mutate_rows call has unfinished work.
+    """
+
+    pass
+
+
+async def _mutate_rows_operation(
+    gapic_client: "BigtableAsyncClient",
+    table: "Table",
+    mutation_entries: list["RowMutationEntry"],
+    operation_timeout: float,
+    per_request_timeout: float | None,
+):
+    """
+    Helper function for managing a single mutate_rows operation, end-to-end.
+
+    Args:
+      - gapic_client: the client to use for the mutate_rows call
+      - table: the table associated with the request
+      - mutation_entries: a list of RowMutationEntry objects to send to the server
+      - operation_timeout: the timeout to use for the entire operation, in seconds.
+      - per_request_timeout: the timeout to use for each mutate_rows attempt, in seconds.
+          If not specified, the request will run until operation_timeout is reached.
+    """
+
+    predicate = retries.if_exception_type(
+        core_exceptions.DeadlineExceeded,
+        core_exceptions.ServiceUnavailable,
+        _MutateRowsIncomplete,
+    )
+
+    retry = retries.AsyncRetry(
+        predicate=predicate,
+        timeout=operation_timeout,
+        initial=0.01,
+        multiplier=2,
+        maximum=60,
+    )
+    # use generator to lower per-attempt timeout as we approach operation_timeout deadline
+    attempt_timeout_gen = _attempt_timeout_generator(
+        per_request_timeout, operation_timeout
+    )
+    # create partial function to pass to trigger rpc call
+    metadata = _make_metadata(table.table_name, table.app_profile_id)
+    gapic_fn = functools.partial(
+        gapic_client.mutate_rows,
+        table_name=table.table_name,
+        app_profile_id=table.app_profile_id,
+        metadata=metadata,
+    )
+
+    # wrap attempt in retry logic
+    attempt_context = _MutateRowsAttemptContext(
+        gapic_fn,
+        mutation_entries,
+        attempt_timeout_gen,
+        predicate,
+    )
+    retry_wrapped = retry(attempt_context.run_attempt)
+    # convert RetryErrors from retry wrapper into DeadlineExceeded errors
+    deadline_wrapped = _convert_retry_deadline(retry_wrapped, operation_timeout)
+    try:
+        # trigger mutate_rows
+        await deadline_wrapped()
+    except Exception as exc:
+        # exceptions raised by retryable are added to the list of exceptions for all unfinalized mutations
+        for idx in attempt_context.remaining_indices:
+            attempt_context.append_error(idx, exc)
```

Comment:
I think it might be better to move this logic into the attempt_context wrapped in a method called finalize_state() or something similar...it will limit the logic leak